packages/worker/src/isolate-builder.ts

```diff
@@ -0,0 +1,151 @@
+    try {
+      const prebuildDir = path.join(distDir, 'prebuild');
+      await fs.mkdir(prebuildDir, { recursive: true });
+      await this.registerActivities(prebuildDir);
+      const entrypointPath = path.join(prebuildDir, 'main.js');
+      const workflows = await this.findWorkflows();
+      await this.genEntrypoint(entrypointPath, workflows);
+      await this.bundle(entrypointPath, distDir);
+      const code = await fs.readFile(path.join(distDir, 'main.js'), 'utf8');
+      const snapshot = ivm.Isolate.createSnapshot([{ code }]);
+      return new ivm.Isolate({ snapshot });
+    } finally {
+      await fs.remove(distDir);
+    }
+  }
+
+  protected async findWorkflows(): Promise<string[]> {
+    const files = await fs.readdir(this.workflowsPath);
+    return files.filter((f) => path.extname(f) === '.js').map((f) => path.basename(f, path.extname(f)));
+  }
+
+  protected async genEntrypoint(target: string, workflows: string[]): Promise<void> {
+    let code = dedent`
+      const init = require('@temporalio/workflow/lib/init');
+      const internals = require('@temporalio/workflow/lib/internals');
+      internals.state.activityDefaults = ${JSON.stringify(this.activityDefaults)};
+
+      const workflows = {};
+      export { init, internals, workflows };
+    `;
+    for (const wf of workflows) {
+      code += `workflows[${JSON.stringify(wf)}] = require(${JSON.stringify(path.join(this.workflowsPath, wf))});\n`;
+    }
+    await fs.writeFile(target, code);
+  }
+
+  protected async bundle(entry: string, distDir: string): Promise<void> {
+    await new Promise<void>((resolve, reject) =>
+      webpack(
+        {
+          resolve: {
+            modules: [this.nodeModulesPath],
+            extensions: ['.js'],
+            alias: Object.fromEntries(
+              [...this.activities.keys()].map((spec) => [spec, path.resolve(distDir, 'prebuild', spec)])
+            ),
+          },
+          entry: [entry],
+          mode: 'development',
+          output: {
+            path: distDir,
+            filename: 'main.js',
+            library: 'lib',
+          },
+        },
+        (err, stats) => {
+          const lines = stats.toString({ chunks: false, colors: true }).split('\n');
+          for (const line of lines) {
+            this.logger.info(line);
+          }
+          if (err) {
+            reject(err);
+          } else if (stats.hasErrors()) {
+            reject(new Error('Webpack stats has errors'));
+          } else {
+            resolve();
+          }
+        }
+      )
+    );
+  }
+
+  public async registerActivities(prebuildDir: string): Promise<void> {
+    validateActivityOptions(this.activityDefaults);
+    const serializedOptions = JSON.stringify(this.activityDefaults);
+    for (const [specifier, module] of this.activities.entries()) {
+      let code = dedent`
+        import { scheduleActivity } from '@temporalio/workflow';
```

Comment:
Not super important, but maybe worth considering using https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#user-content-creating-and-printing-a-typescript-ast instead of constructing JS by creating strings. A more type safe way to codegen.